Split `RenderingDevice` into API-agnostic and `RenderingDeviceDriver` parts

RandomShaper opened the pull request:
A huge chunk of `RenderingDeviceVulkan` is API-agnostic bookkeeping, validation, etc. Such code is also shared with others, like `RenderingDeviceD3D12`. This PR keeps a single, non-abstract `RenderingDevice` with all those API-agnostic elements and moves all the API-specific parts to a thin wrapper around the driver, which is `RenderingDeviceDriver` for Vulkan. This makes much easier to write additional rendering drivers, by allowing one to focus on the API-specific stuff. In some cases that will require modifying the interface between `RD` and `RDD`, which is not exposed to the user.

Other relevant points:
- A common context interface has been extracted from `VulkanContext` (`APIContextRD`) with the bare minimum set of services `RenderingDevice` and direct usage needs. (The specific `RenderingDeviceDriver`s have access to the specific context implementation, which open doors for them to a much richer, but non-standard, API.) **UPDATE:** Now Direct3D 12 has been merged, it gets similar love in this PR.
- There are important notes on design choice in the code. Future maintainers, please follow or discuss, but don't break consistency.
- Every struct involved now have default initializers, both in `RenderingDevice` and `RenderingDeviceDriver`.
- Some of the wiring between context and `RD` has been enhanced so now you can instantiate a `RenderingDevice` even when Godot is running in compatibility mode (this will allow to have a `RD`-based lightmapper available in a GL compatibility project). This piece of code shows how to ensure you have a usable local `RenderingDevice` in any circumstances:
   ```C++
   RD *orig_rd_singleton = RD::get_singleton();

   ApiContextRD *vc = nullptr;
   if (GLOBAL_GET("rendering/renderer/rendering_method") == "gl_compatibility") {
   #ifdef VULKAN_ENABLED
   	if (GLOBAL_GET("rendering/rendering_device/driver") == "vulkan") {
   		vc = memnew(VulkanContext);
	}
   #endif
   	CRASH_COND_MSG(!vc, "No RD driver available.");
   	vc->initialize();
   } else {
   	vc = RD::get_singleton()->get_context();
   }

   RenderingDevice *rd = memnew(RenderingDevice);
   if (GLOBAL_GET("rendering/renderer/rendering_method") == "gl_compatibility") {
   	DEV_ASSERT(RD::get_singleton() == rd);
   }
   rd->initialize(vc, true);
   // Do stuff with RD here.
   memdelete(rd);

   if (GLOBAL_GET("rendering/renderer/rendering_method") == "gl_compatibility") {
    	memdelete(vc);
   }
   DEV_ASSERT(RD::get_singleton() == orig_rd_singleton);
   ```
- Deprecations: `RD::texture_get_native_handle()` (superfluous) and `DRIVER_RESOURCE_VULKAN_*` (replaced by API-agnostic).

~This includes a couple of merged (commits will go away once rebased) and unmerged (#82797) PRs.~

TODO:
- [X] Finish debugging and fix a few known issues.
- [X] Rebase on top of the lastest `master`.
- [X] Fix ligthmapper (local device?).
- [X] D3D12: Adapt to this new infrastucture. **2023-12-11:** The PR now includes #70315, which is then adapted to the new model.
- [X] Debug
- [X] Remove the interim change-tracking-assistance commit.
- [X] Ensure #86106 is fixed.

_Production edit: closes godotengine/godot-roadmap#31_

reduz commented on servers/rendering/rendering_device.h:
```diff
@@ -50,56 +58,16 @@ class RDPipelineColorBlendState;
 class RDPipelineSpecializationConstant;
 
-class RenderingDevice : public Object {
+class RenderingDevice : public Object, public RenderingDeviceCommons {
```
This is probably not going to work (or not warranted to work) with Godot binding system because depending on the implementation of multiple inheritance, the compiler will do offset adjustments to the pointer when casting, which breaks Godot type system.

This is why I originally suggested doing this different.

reduz replied:
Alternatively you could make RenderingDeviceCommons inherit from object and simply not registering RenderingDeviceDriver to the binding system (no GDCLASS) that should work fine.

RandomShaper replied:
I'll do just that in the next push.

reduz commented on servers/rendering/renderer_rd/api_context_rd.h:
```diff
@@ -0,0 +1,77 @@
+	virtual RDD::FramebufferID window_get_framebuffer(DisplayServer::WindowID p_window = 0) = 0;
+
+	virtual RID local_device_create() = 0;
+	virtual void local_device_push_command_buffers(RID p_local_device, const RDD::CommandBufferID *p_buffers, int p_count) = 0;
```
This will probably need to be changed for the acyclic graph, exposing queues and synchronization primitives.

RandomShaper replied:
For now, I've chosen to alter the design of the context class as little as possible to keep the scope of the work under control for this first big iteration, but I won't deny future maintainers will need to make this more flexible (probably exposing from the driver).

reduz commented:
Looks fantastic!

DarioSamo commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+	};
+
+	struct AttachmentReference {
+		static const uint32_t UNUSED = 0xffffffff;
```
This constant seems like it'd be better off defined at the global level, as the code below uses this same constant but without a name.

reduz replied:
IMO this should go away as we remove subpasses anyway.

RandomShaper replied:
The one below is for subpass, and is just a garbage, but fixed default value to make it apparent when it's not properly filled. This one is a proper value that can be specified as an (unused) attachment index.

DarioSamo commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+		uint32_t layer_count = 0xffffffff;
+	};
+
+	virtual void command_begin_render_pass(CommandBufferID p_cmd_buffer, RenderPassID p_render_pass, FramebufferID p_framebuffer, CommandBufferType p_cmd_buffer_type, const Rect2i &p_rect, VectorView<RenderPassClearValue> p_clear_values) = 0;
```
Minor nitpick, what's the benefit of CommandBufferType being specified here? Shouldn't we be able to derive it from the command buffer itself? Is it a validation measure?

reduz replied:
Yeah thats a good point. That said I think we may want to simplify this.

RandomShaper replied:
Well, since `RD` knows if it's about a primary or secondary command buffer, that argument avoids the need for `RDD` to track it and be able to map `RD::CommandBufferID` to `VkCommandBuffer` directly. The parameter is leveraged this way:
```
	VkSubpassContents vk_subpass_contents = p_cmd_buffer_type == COMMAND_BUFFER_TYPE_PRIMARY ? VK_SUBPASS_CONTENTS_INLINE : VK_SUBPASS_CONTENTS_SECONDARY_COMMAND_BUFFERS;
	vkCmdBeginRenderPass((VkCommandBuffer)p_cmd_buffer.id, &render_pass_begin, vk_subpass_contents);
```

I'm not sure how orthogonal are command buffer level and subpass content level as concepts, but the fact is that currently `RD` uses them 1:1, so I just took advantage.

DarioSamo commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+	/**** SCREEN ****/
+	/****************/
+
+	virtual DataFormat screen_get_format() = 0;
```
This is a weird concept, as there's nothing exposed about how swap chains work or for what device they're currently associated to. You don't exactly tie a 'driver/device' to a single screen either. Probably one of those quirks we need to work on when Vulkan Context is re-evaluated.

RandomShaper replied:
Yeah, the boundary between context and device is kind of blurry. The design warrants a good redesign pass.

DarioSamo commented:
I'll be leaving feedback in general here and on a per-line basis if necessary. Some of my comments are just criticizing a few design carry-overs from the current design and aren't necessarily related to this PR being approved or not, it just feels like the right time to bring these points up.

### Type Safety

I already brought this up over DMs, but there's currently no type safety system for the IDs returned by RDD. Using the uint64_t's directly means we can mix and match arguments around and the compile won't indicate the error. We can solve this in a few ways.

**A)** Every type returned gets a struct that wraps around the uint64_t. A very minor change that guarantees the types can't be mixed together.
```
struct FramebufferID {
  uint64_t handle;
};
```

**B)** Every type returned is a pointer to an interface that the backend has to inherit. This requires more work but since these handles are already pointers (they're not references you can indefinitely hold) in disguise, there's not much of a downside to it. The major advantage is that debuggers can access these types natively and display their contents.
```
struct Framebuffer {
  // Empty interface.
};

virtual Framebuffer *framebuffer_create() = 0;
virtual void framebuffer_free(Framebuffer *p_framebuffer) = 0;
```
```
struct FramebufferVK : Framebuffer {
  VkFramebuffer fb = VK_NULL_HANDLE;
}
```
Another benefit to consider is const-correctness can be implemented into the pointer types and will therefore be directly indicated by the API whether the object's internals should be modifiable or not.

### Vertex Formats

There seems to be currently a forced correlation between bindings and attributes, meaning that attributes that share the same buffer can't share the same binding (e.g. if 10 elements are bound to the same buffer, they each get their own binding instead of using the same one). I'm not familiar if this has potential performance implications, but it's worth investigating if this is correct.

This is not new to this PR, as it was pretty much directly copied as is from the current Vulkan RD.

### Command Queues

Command queues are not directly exposed but managed by the backend, meaning RenderingDevice has no control whatsoever on where commands are executed, on what type of queue and when exactly. This could prove to be very useful in the feature as a lot of the logic around managing command queues should be common to most backends and it takes off some responsibility for the backend maintainer. I think we could address this in this PR and move most of the logic that Vulkan Context currently handles to RD instead.

I'll do a separate comment for concerns about D3D12 compatibility in the future.

reduz commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+	virtual void buffer_free(BufferID p_buffer) = 0;
+	virtual uint64_t buffer_get_allocation_size(BufferID p_buffer) = 0;
+	virtual uint8_t *buffer_map(BufferID p_buffer) = 0;
+	virtual void buffer_unmap(BufferID p_buffer) = 0;
```
Probably better to keep it simple, I don't think its needed.

RandomShaper replied:
As with other features in the driver, I've generally chosen to expose what is currently actually used instead of opening dorrs to the whole functionality. This is something that future maintainers (maybe yourself) will have to change in case there's an actual use case for partial mapping.

stuartcarnie replied:
For reference, Metal doesn't have to map / unmap buffers. `MTLBuffer` just has a pointer to the buffer, and you must respect standard memory coherency rules.

reduz commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+
+	virtual String shader_get_binary_cache_key() = 0;
+	virtual Vector<uint8_t> shader_compile_binary_from_spirv(VectorView<ShaderStageSPIRVData> p_spirv, const ShaderReflection &p_shader_refl, const String &p_shader_name) = 0;
+	virtual ShaderID shader_create_from_bytecode(const Vector<uint8_t> &p_shader_binary, ShaderDescription &r_shader_desc, String &r_name) = 0;
```
Binary is implementation dependent. If you use Vulkan its likely still SPIRV, but if you use D3D12 or Metal, its the native format.

reduz commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+	/****************/
+
+	virtual String shader_get_binary_cache_key() = 0;
+	virtual Vector<uint8_t> shader_compile_binary_from_spirv(VectorView<ShaderStageSPIRVData> p_spirv, const ShaderReflection &p_shader_refl, const String &p_shader_name) = 0;
```
So far this is the responsibility of the backend, because it returns the format that is best preferred for caching in the driver for binary.

That said, when exporting we may want to ship caches precompiled, so these compilation things may need to be able from somewhere else, but at the moment it does not matter and I would leave as is. That's Pedro's problem when he does this export optimization.

RandomShaper replied:
For the time being, yes, everything orbits around SPIR-V. That said, the part that is API-agnostic, which is the reflection into Godot `RenderingDevice` structures is performed by `RenderingDevice`. From that info, the driver has to put it into it's own format. The D3D12 backend does the translation to DXIL as part of that process. Honestly, other parts of the process of writing the Godot shader binary, like compressing the bytecode of the stages, etc. is very similar or identical between Vulkan and D3D12, and can probably be refactored into common and distinct parts... in a future iteration.

reduz commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+		ATTACHMENT_LOAD_OP_LOAD = 0,
+		ATTACHMENT_LOAD_OP_CLEAR = 1,
+		ATTACHMENT_LOAD_OP_DONT_CARE = 2,
+		ATTACHMENT_LOAD_OP_NONE = 1000400000,
```
Yeah possibly not needed.

RandomShaper replied:
I'm removing them (also the store counterpart).

reduz commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+	enum AttachmentStoreOp {
+		ATTACHMENT_STORE_OP_STORE = 0,
+		ATTACHMENT_STORE_OP_DONT_CARE = 1,
+		ATTACHMENT_STORE_OP_NONE = 1000301000,
```
agreed I think its not needed, is it used?

RandomShaper replied:
It's not. I'll remove it.

RandomShaper commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+	};
+
+	struct SubpassDependency {
+		uint32_t src_subpass = 0xffffffff;
```
See above. This is just a fixed, but garbage default.

DarioSamo commented:
These are not something that necessarily need to be solved by this PR but are probably worth considering before doing the D3D12 version of RDD. Some are really just points to consider when doing the D3D12 refactor.

### Combined Image Samplers

As far as I know these are not supported by D3D12 and the backend would need to work around it in some ugly ways to support them. Are we relying on this feature a lot? What approach did the D3D12 PR take to solving this?

### Compute dispatch base

I'm not aware D3D12 has anything about this. This is a new addition to the API by this PR. We might want to consider it more carefully. It's also a Vulkan 1.1 feature at minimum, so we need to query device support for it. I think we should just remove it and let the user handle it by adding the offset in a UBO/Push constant.

### Barriers

D3D12 requires resource barriers if not using enhanced barriers. I think we should take the chance to make our barrier-API only take resources instead. I don't think we ever rely on memory barriers. As far as I'm aware the current D3D12 PR needs to work around this and I think it's hard to find use cases for full memory barriers with the upcoming acyclic render graph/barrier solver design.

### Semantics vs Locations

What approach does the current D3D12 PR take for matching vertex location inputs to the semantics? Is that something that we'd need to address on this API or is it solved through some other means as part of the backend responsibility? If there's a possibility this API will take shaders in its native format, it sounds like we might need to allow for a type of configuration that allows input semantics.

### Unordered Access State

What is the suggested solution for mapping this API to the unordered access state in D3D12? Using the image layout for general and the bit for shader writes?

### Texel Buffers

These don't translate to a flag during creation in D3D12, so we'll just need to track that bit to enable using the format on the SRV/UAV when it's assigned.

RandomShaper commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,619 @@
+	virtual bool pipeline_cache_create(const Vector<uint8_t> &p_data) = 0;
+	virtual void pipeline_cache_free() = 0;
+	virtual size_t pipeline_cache_query_size() = 0;
+	virtual Vector<Vector<uint8_t>> pipeline_cache_serialize() = 0;
```
I'm not very concerned about the allocations, since, after all, that's being called when about to hit the disk. Also, one of the chunks has actually been already allocated by the driver.

Regarding the obscurity of the return type, I can't think of a good one. It's just a number of vectors the caller is expected to write back-to-back in the cache file.

DarioSamo replied:
Why is it a Vector of Vectors but the input is a Vector though? Am I supposed to concatenate them and then use them back again as the input?

RandomShaper replied:
That _was_ the idea. And I'm saying _was_ because on closer inspection I've realized I can get rid of that ugliness. Next push will come with a pain `Vector<uint8_t>` as the return type.

RandomShaper commented:
You may want to check the D3D12 PR (#70315) to see how I mapped the `RenderingDevice` features to it.

Regarding barriers, the D3D12 driver using resource-level barriers would just ignore the detailed ones (that's exactly what the PR above is doing). If enhanced barriers are eventually implemented, well, it just have to obey them.

I'm addressing the type safey of ids in the next push.

Regarding other things, this is just the first iteration and will for sure need changes to fit other APIs or future `RD` features better.

RandomShaper commented:
Regarding type safety, I'm taking the first approach you discuss because most `RDD` ids aren't actually pointers to its view of resources, but just the Vulkan handles.

stuartcarnie commented on servers/rendering/rendering_device_commons.cpp:
Excellent – moving these APIs to a common place so they don't have to be duplicated a 3rd time (Vulkan, D3D12 and Metal) is 💯

stuartcarnie commented on servers/rendering/rendering_device.cpp:
```diff
@@ -91,771 +442,6324 @@ RID RenderingDevice::shader_create_from_spirv(const Vector<ShaderStageSPIRVData>
+}
+
+Vector<uint8_t> RenderingDevice::shader_compile_binary_from_spirv(const Vector<ShaderStageSPIRVData> &p_spirv, const String &p_shader_name) {
+	ShaderReflection reflection;
+	if (_reflect_spirv(p_spirv, reflection) != OK) {
+		return Vector<uint8_t>();
+	}
```
I need to use SPIRV-Cross for Metal, in order to generate MSL from the SPIR-V. I realise the `_reflect_spirv` does some validation, but otherwise, I don't really need it, so it is wasted CPU.

Does it make sense to leave it to the driver to decide if it wants to use `_reflect_spirv`?

RandomShaper replied:
Very good point. In the next push, I'm making `_reflect_spirv()` an optional service that drivers can choose to use or not.

stuartcarnie replied:
Disregard, as I see I'll use this as a sanity chec

RandomShaper committed:
BitField: allow chained set_flag(), make constructors constexpr

RandomShaper commented:
Re-pushed. Lightmapper is fixed (or so I think).

RandomShaper committed:
RBMap: Add explicit copy operators to iterators

Absence thereof is deprecated and breaks builds on most compilers.

Bonus: Fix parameter naming style throughout.

RandomShaper commented:
@sakrel, thank you for your feedback. I acknowledge that the new single `RenderingDevice` class in this PR is indeed tailored to the current supported, and about to be supported, Vulkan and D3D12. Therefore, it's indeed unsuitable for APIs following different models.

The best option I can think of to make WebGPU fit again is just making `RenderingDevice` more flexible, in collaboration with the driver. As such, `RenderingDeviceDriver` would have a new function to query about the model of the underlying API-platform. `RenderingDevice` would get extra logic to handle the different possibilities. `RenderingDeviceDriver` would, in turn, have to provide new functions (I think) for the pieces of functionality that need different mechanics (e.g., callback system for async), that would only be implemented for `RenderingDeviceDriverWebGPU`, and any other potentially needing similar characteristics in the future.

As I see it , the interface of `RenderingDeviceDriver` will be subject to change for a long time (potentially forever), to let it adapt to any needs that arise on the side of `RenderingDevice`, to support all its present and future features, as well as every graphics API.

RandomShaper commented on core/templates/rb_map.h:
```diff
@@ -116,6 +116,11 @@ class RBMap {
 			return E != nullptr;
 		}
+
+		Iterator &operator=(const Iterator &b) {
+			E = b.E;
```
In this kind of cases, I'd rather keep the PR consistent with the _de facto_ style in the surroundings, and maybe then make another one fixing the whole file. That's why I've followed a noncompliant style here.

RandomShaper replied:
Fair enough. I'm amending the relevant commit with `p_it` in every case, as a bonus. Now I'm wondering about the usage of `p_other` vs. `p_type_in_snake_case` in special member functions across the whole codebase. Not super important (and not something for now), but the greater the consistency, the merrier. 😃

BastiaanOlij commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -0,0 +1,3247 @@
+
+	textures.insert(vk_image_view, ti);
+
+	return TextureID(vk_image);
```
This needs to be `return TextureID(vk_image_view);` just like the other create functions.

RandomShaper replied:
Done! Will come in the next push.

BastiaanOlij commented:
After applying the fix I mentioned in my comment I tested this with https://github.com/BastiaanOlij/godot_openxr_demo and it now works for OpenXR.

Tested using the mobile renderer with:
- Valve Index running in PCVR mode
- Quest Pro running natively
- Quest 3 running natively (surprising, works without the bypass implemented in #84244, did you do something to cause presentation to be skipped when there is nothing to present?)

Performance seems on-par with Godot 4.2 stable, not surprising as the mobile renderer likely has less to gain from these improvements.

atirut-w commented:
> Performance seems on-par with Godot 4.2 stable

Would this make a difference when multiple render passes is implemented? I feel like this would have a more impact once it comes to thatl

BastiaanOlij commented:
@RandomShaper sadly I found a more serious issue. Things go really bad when I enable MSAA.

On desktop it's rendering fine, but inside the Quest I'm getting all sorts of issues with polygons flickering. It very much looks like that it is either resolving the buffer before its finished rendering, or it may be an issue with a missing barrier before the results are handed off to the XR Compositor (seeing this happens outside of the control of the rendering device, it has likely no idea a barrier is needed).

Interestingly enough recording, while the issues remain visible in headset, are not present in the recording, which suggest Meta might introduce a barrier here.

I suspect that I didn't have trouble on PCVR simply because we finish rendering WAY before the XR compositor decides to present the image, I'm at 0.5ms render time so the GPU isn't even breaking a sweat.

DarioSamo commented:
> > Performance seems on-par with Godot 4.2 stable
> 
> Would this make a difference when multiple render passes is eventually implemented? I feel like this would have a more impact once it comes to that.

RDD isn't expected to have performance improvements on its own as far as I know since the barrier management is the same as what it was before.

YuriSizov commented:
Just an update from the production team. We're ready to merge this whenever it's ready, it's next in the line of big rendering changes. We would be okay merging it for dev1, which we plan to have early next week. But if you'd rather have it merged afterward dev1, that's also okay. The next public build will only be made in January, so that won't yield any broad user testing yet, but at least your burden with juggling branches will be easier.

With that said, this still needs a rebase from @RandomShaper to finalize the commits and remove the D3D12 prerequisites.
And this still needs some approvals and positive reviews to formally proclaim it as ready. So cc @clayjohn and @BastiaanOlij. And I'm ready when you all are. 🙃

DarioSamo commented:
I can attest I've been doing some very extensive testing of this branch as well as people who've tested the Acyclic Graph have been using this. I'll be doing a rebase when the final version of this PR is submitted to double-check everything still works as intended.

RandomShaper commented:
@YuriSizov, thanks for the heads-up. The problem is that,  now that D3D12 is merged, this PR is not complete until I'm done bug fixing the RDD incarnation of it. So, this can only be put in dev1 if I can do that in time. (Likely and desirable.)

YuriSizov commented:
@RandomShaper No pressure from us in that regard 🙂

BastiaanOlij commented:
@YuriSizov @RandomShaper my testing so far with this, it all looks pretty good.

My only issue I ran into was on Quest where we seem to have a sync or fence issue introduced. The Quest starts consuming the image before we finish rendering it. @DarioSamo and I were looking into that and I've still got some tests to complete.

However I think that shouldn't be a showstopper for this to go ahead esp as we're recommending everyone on Quest to use compatibility already. It's something we can fix up later.

RandomShaper committed:
Split RenderingDevice into API-agnostic and RenderingDeviceDriver parts

Credit and thanks to @bruzvg for multiple build fixes, update of 3rd-party items and MinGW support.

Co-authored-by: bruvzg <7645683+bruvzg@users.noreply.github.com>

bruvzg commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -0,0 +1,663 @@
+class RenderingDeviceDriver : public RenderingDeviceCommons {
+public:
+	struct ID {
+		uint64_t id = 0;
+		_ALWAYS_INLINE_ ID() = default;
+		_ALWAYS_INLINE_ ID(uint64_t p_id) :
+				id(p_id) {}
```
I guess this should be changed to `size_t` to work with 32-bit builds (static assert below).

RandomShaper replied:
Good point. Poor 32-bit builds, always forgotten.

bruvzg commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -0,0 +1,3347 @@
+uint64_t RenderingDeviceDriverVulkan::api_trait_get(ApiTrait p_trait) {
+	switch (p_trait) {
+		case API_TRAIT_TEXTURE_TRANSFER_ALIGNMENT:
+			return (uint64_t)MAX(16, limits.optimalBufferCopyOffsetAlignment);
```
```suggestion
			return (uint64_t)MAX(16ULL, limits.optimalBufferCopyOffsetAlignment);
```

reduz reviewed (approved):
Looks fine on my end, incredible work Pedro!

RandomShaper commented on core/config/project_settings.cpp:
```diff
@@ -1399,12 +1399,22 @@ ProjectSettings::ProjectSettings() {
 	GLOBAL_DEF("rendering/rendering_device/pipeline_cache/save_chunk_size_mb", 3.0);
 	GLOBAL_DEF("rendering/rendering_device/vulkan/max_descriptors_per_pool", 64);
+
+#ifdef D3D12_ENABLED
```
Yeah, sounds like the most reasonable thing to do.

clayjohn commented:
Before jumping into code review. Just a few notes on the testing this branch has had (that I am aware of and apart from the testing that is mentioned above):
1. Used in https://github.com/godotengine/godot/pull/84976 
2. Used by stuart carnie in developing the Metal backend
3. Used by The Forge as their base for Vulkan optimizations
4. Tested by me in multiple scenes of varying complexity 
    a. Test scene from https://www.youtube.com/watch?v=9kcjFJJxO6I&t=1484s (tons of assets, textures, materials, GI, heavy shaders)
    b. TPS demo with different permutations of GI, fog, etc.
    c. Simplfied TPS demo using mobile renderer, but with heavy use of animations, particles, and decals
    
At this point any problems that have been identified have been fixed and are included in the current state of things. In testing the current iteration I did not identify any issues (no graphical bugs, no crashes, no new errors, no new validation layer errors)

clayjohn reviewed (approved):
The code looks good to me. I can't spot anything that I would want changed before merging. There is some discussion above about XR issues that may have been exposed by this PR. Bastiaan says he is okay fixing those after merging and I agree. It is very likely that those issues will be fixed by https://github.com/godotengine/godot/pull/84976 anyway which we will be merging shortly. 

I think it is time to move ahead with this PR and merge it!

bruvzg reviewed (approved):
Tested on macOS and iOS (rebased on top of master, since https://github.com/godotengine/godot/pull/86219 is required). Seems to work fine.

DX12 version on Windows also seems to work.

dsnopek commented:
Because I saw @BastiaanOlij's comments about this not working on the Quest, I decided to try it myself, and, unfortunately, and this PR seems to completely break the Compatibility renderer. :-/ When I switch to the Vulkan Mobile renderer, it seems to work fine. So, kind of the reverse of what of what Bastiaan saw?

This is the error I'm getting with the Compatibility renderer (via `adb logcat`):

```
12-19 13:59:25.549  4418  4459 E godot   : USER ERROR: Parameter "native_window" is null.
12-19 13:59:25.549  4418  4459 E godot   :    at: DisplayServerAndroid (platform/android/display_server_android.cpp:572)
12-19 13:59:25.549  4418  4459 E godot   : USER ERROR: Unable to create DisplayServer, all display drivers failed.
12-19 13:59:25.549  4418  4459 E godot   :    at: setup2 (main/main.cpp:2466)
12-19 13:59:25.549  4418  4459 E godot   : USER ERROR: Condition "!_start_success" is true. Returning: false
12-19 13:59:25.549  4418  4459 E godot   :    at: start (main/main.cpp:2928)
```

And that last message is spammed over and over.

I suspect there's something slightly off in the Android changes?

bruvzg commented:
> I suspect there's something slightly off in the Android changes?

```
ANativeWindow *OS_Android::get_native_window() const {
#if defined(VULKAN_ENABLED)
	return native_window;
#else
	return nullptr;
#endif
}
```

So these lines should be inside Vulkan check.
```
		ANativeWindow *native_window = OS_Android::get_singleton()->get_native_window();
		ERR_FAIL_NULL(native_window);
```

bruvzg commented on platform/android/display_server_android.cpp:
```diff
@@ -518,20 +517,31 @@ void DisplayServerAndroid::register_android_driver() {
+#endif
+		} wpd;
+#ifdef VULKAN_ENABLED
+		if (rendering_driver == "vulkan") {
+			wpd.vulkan.window = native_window;
+		}
```
```suggestion
		if (rendering_driver == "vulkan") {
			ANativeWindow *native_window = OS_Android::get_singleton()->get_native_window();
			ERR_FAIL_NULL(native_window);
			wpd.vulkan.window = native_window;
		}
```

bruvzg commented on platform/android/display_server_android.cpp:
```diff
@@ -518,20 +517,31 @@ void DisplayServerAndroid::register_android_driver() {
-	if (rendering_driver == "vulkan") {
+#if defined(RD_ENABLED)
+	if (context_rd) {
 		ANativeWindow *native_window = OS_Android::get_singleton()->get_native_window();
 		ERR_FAIL_NULL(native_window);
```
```suggestion
```

bruvzg commented on platform/android/display_server_android.cpp:
```diff
@@ -554,30 +564,47 @@ DisplayServerAndroid::DisplayServerAndroid(const String &p_rendering_driver, Dis
+	context_rd = nullptr;
+	rendering_device = nullptr;
+
+	ANativeWindow *native_window = OS_Android::get_singleton()->get_native_window();
+	ERR_FAIL_NULL(native_window);
```
```suggestion
```

bruvzg commented on platform/android/display_server_android.cpp:
```diff
@@ -554,30 +564,47 @@ DisplayServerAndroid::DisplayServerAndroid(const String &p_rendering_driver, Dis
+		} wpd;
+#ifdef VULKAN_ENABLED
+		if (rendering_driver == "vulkan") {
+			wpd.vulkan.window = native_window;
```
```suggestion
			ANativeWindow *native_window = OS_Android::get_singleton()->get_native_window();
			ERR_FAIL_NULL(native_window);
			wpd.vulkan.window = native_window;
```

dsnopek commented on platform/android/display_server_android.cpp:
```diff
@@ -554,30 +564,47 @@ DisplayServerAndroid::DisplayServerAndroid(const String &p_rendering_driver, Dis
+#if defined(VULKAN_ENABLED)
 	if (rendering_driver == "vulkan") {
-		ANativeWindow *native_window = OS_Android::get_singleton()->get_native_window();
-		ERR_FAIL_NULL(native_window);
```
This change seems to be the source of my problems when using the Compatibility renderer. Previously, the `ERR_FAIL_NULL(native_window);` would only happen when using Vulkan, but this PR causes it to happen always.

Applying this diff fixes it for me:

```
diff --git a/platform/android/display_server_android.cpp b/platform/android/display_server_android.cpp
index a133cd3ee2..640441a4b3 100644
--- a/platform/android/display_server_android.cpp
+++ b/platform/android/display_server_android.cpp
@@ -568,9 +568,6 @@ DisplayServerAndroid::DisplayServerAndroid(const String &p_rendering_driver, Dis
        context_rd = nullptr;
        rendering_device = nullptr;
 
-       ANativeWindow *native_window = OS_Android::get_singleton()->get_native_window();
-       ERR_FAIL_NULL(native_window);
-
 #if defined(VULKAN_ENABLED)
        if (rendering_driver == "vulkan") {
                context_rd = memnew(VulkanContextAndroid);
@@ -578,6 +575,9 @@ DisplayServerAndroid::DisplayServerAndroid(const String &p_rendering_driver, Dis
 #endif
 
        if (context_rd) {
+               ANativeWindow *native_window = OS_Android::get_singleton()->get_native_window();
+               ERR_FAIL_NULL(native_window);
+
                if (context_rd->initialize() != OK) {
                        memdelete(context_rd);
                        context_rd = nullptr;
```

RandomShaper replied:
Makes sense! The fix will come in the imminent next push.

RandomShaper committed:
Make PagedAllocator more compatible (esp., with HashMap)

DarioSamo commented:
I found a particular project that doesn't work with this branch in D3D12 mode, I'll see if I can get an MRP once I narrow down what the project does that is incompatible, but it seems it's straight up unable to create the graphics PSOs on this project.

RandomShaper commented:
Pushed removing many cases of runtime allocation (powered by `PagedAllocator`), plus a change to `PagedAllocator` to improve its compatibility (e.g., usable with `HashMap`), plus the fix discussed above.

> > I suspect there's something slightly off in the Android changes?
> 
> ```
> ANativeWindow *OS_Android::get_native_window() const {
> #if defined(VULKAN_ENABLED)
> 	return native_window;
> #else
> 	return nullptr;
> #endif
> }
> ```
> 
> So these lines should be inside Vulkan check.
> 
> ```
> 		ANativeWindow *native_window = OS_Android::get_singleton()->get_native_window();
> 		ERR_FAIL_NULL(native_window);
> ```

I already fixed it locally, but hadn't pushed yet.

RandomShaper commented:
I've pushed with the requested fixes, plus a small change to use `PagedAllocator` for the multiple RDD-bookkept resources, instead of using plain `memnew()`.

Regarding the SDFGI issue reported by @DarioSamo, it's a known issue the pre-RDD Direct3D 12 renderer already had. (See _Deprecate Texture Aliasing_ in #70315.) It will be addressed separately.

RandomShaper commented:
News on the SDFGI thing: there was a silly bug initializing textures requiring format sharing. On Windows 11 you don't even need the Agility SDK for the feature to work without the hack, so I've pushed a fix to let it work. (The hacky version still works for me on W10 with no Agility SDK, on Nvidia. A proper fix for other GPUs will be provided in the future.)

RandomShaper commented:
If CI succeeds (it failed because of a C++ technicality in the just added `VersatileResource` helper for allocation avoidance).

dsnopek reviewed (approved):
Thanks! I'm not qualified to review the RD changes (aka the main point of this PR) but I just retested the latest with the Compatibility renderer on Quest 3, and it's working now. :-)

YuriSizov commented:
Okay, so Android/Quest issues have been addressed, and if the CI passes we're ready to go. Correct? 🙂

DarioSamo commented:
> Okay, so Android/Quest issues have been addressed, and if the CI passes we're ready to go. Correct? 🙂

I just reported an issue with the latest change so just a bit more.

DarioSamo commented:
> Not the D3D12 one from above which is also present in master?

There's just an implementation error in the recent change that will cause the list to get filled incorrectly with the size of the array is > 1. Should be fixed easily.

RandomShaper commented:
It seems to be a GPU driver or OS version specific issue. Works as expected on some setups, fails on another. Maybe a driver bug. We are fine with merging as it is now. The code seems to be OK.

RandomShaper commented:
This is never ending, is it? 😅 Pushed with a fix for the memory leaks.

RandomShaper commented:
I think I'm going to force-push again...

Just kidding. 😈

YuriSizov commented:
Thanks! As often is the case with such achievements, this monumental effort is not the end, but a beginning for even more exciting new things to be added into the engine in near and far future. But tonight you can rest with a proud smile. Cheers to everyone who contributed and tested this PR!

Cyangmou commented:
Really happy to see this, just wanted to say thanks for the effort and can't wait for future improvements this will enable!

aaronfranke commented:
This does not compile on macOS, it fails on the linking stage:

```
[100%] Linking Program bin/godot.macos.editor.arm64 ...
Undefined symbols for architecture arm64:
  "_vkCmdDrawIndexedIndirectCount", referenced from:
      RenderingDeviceDriverVulkan::command_render_draw_indexed_indirect_count(RenderingDeviceDriver::CommandBufferID, RenderingDeviceDriver::BufferID, unsigned long long, RenderingDeviceDriver::BufferID, unsigned long long, unsigned int, unsigned int) in libdrivers.macos.editor.arm64.a(rendering_device_driver_vulkan.macos.editor.arm64.o)
  "_vkCmdDrawIndirectCount", referenced from:
      RenderingDeviceDriverVulkan::command_render_draw_indirect_count(RenderingDeviceDriver::CommandBufferID, RenderingDeviceDriver::BufferID, unsigned long long, RenderingDeviceDriver::BufferID, unsigned long long, unsigned int, unsigned int) in libdrivers.macos.editor.arm64.a(rendering_device_driver_vulkan.macos.editor.arm64.o)
ld: symbol(s) not found for architecture arm64
```

clayjohn commented:
> This does not compile on macOS, it fails on the linking stage (I tested both arm64 and x86_64):
> 
> ```
> [100%] Linking Program bin/godot.macos.editor.arm64 ...
> Undefined symbols for architecture arm64:
>   "_vkCmdDrawIndexedIndirectCount", referenced from:
>       RenderingDeviceDriverVulkan::command_render_draw_indexed_indirect_count(RenderingDeviceDriver::CommandBufferID, RenderingDeviceDriver::BufferID, unsigned long long, RenderingDeviceDriver::BufferID, unsigned long long, unsigned int, unsigned int) in libdrivers.macos.editor.arm64.a(rendering_device_driver_vulkan.macos.editor.arm64.o)
>   "_vkCmdDrawIndirectCount", referenced from:
>       RenderingDeviceDriverVulkan::command_render_draw_indirect_count(RenderingDeviceDriver::CommandBufferID, RenderingDeviceDriver::BufferID, unsigned long long, RenderingDeviceDriver::BufferID, unsigned long long, unsigned int, unsigned int) in libdrivers.macos.editor.arm64.a(rendering_device_driver_vulkan.macos.editor.arm64.o)
> ld: symbol(s) not found for architecture arm64
> ```

It compiles in the CI. Are you using any different parameters that would have an appreciable difference?

Molten VK appears to have support for both those functions, so it is odd that it's not linking

aaronfranke commented:
@clayjohn Fixed, the problem was that I was using an old version of MoltenVK. If anyone else is experiencing this issue, try updating the Vulkan SDK.

AThousandShips commented on platform/linuxbsd/x11/display_server_x11.cpp:
```diff
@@ -5606,10 +5606,21 @@ DisplayServerX11::WindowID DisplayServerX11::_create_window(WindowMode p_mode, V
+			}
+#endif
+			Error err = context_rd->window_create(id, p_vsync_mode, win_rect.size.width, win_rect.size.height, &wpd);
+			ERR_FAIL_COND_V_MSG(err != OK, INVALID_WINDOW_ID, vformat("Can't create a %s window", context_rd->get_api_name()));
```
Related:
* https://github.com/godotengine/godot/issues/86436

akien-mga replied:
Thanks, copied my logs there.
